Mathlib/FieldTheory/PurelyInseparable.lean

```diff
@@ -0,0 +1,579 @@
+
+theorem AlgEquiv.isPurelyInseparable_iff (e : K ≃ₐ[F] E) :
+    IsPurelyInseparable F K ↔ IsPurelyInseparable F E :=
+  ⟨fun _ ↦ e.isPurelyInseparable, fun _ ↦ e.symm.isPurelyInseparable⟩
+
+/-- If `E / F` is an algebraic extension, `F` is separably closed,
+then `E / F` is purely inseparable. -/
+theorem Algebra.IsAlgebraic.isPurelyInseparable_of_isSepClosed (halg : Algebra.IsAlgebraic F E)
+    [IsSepClosed F] : IsPurelyInseparable F E :=
+  ⟨halg.isIntegral, fun x h ↦ minpoly.mem_range_of_degree_eq_one F x <|
+    IsSepClosed.degree_eq_one_of_irreducible F (minpoly.irreducible (halg x).isIntegral) h⟩
+
+variable (F E K)
+
+/-- If `E / F` is both purely inseparable and separable, then `algebraMap F E` is surjective. -/
+theorem IsPurelyInseparable.surjective_algebraMap_of_isSeparable
+    [IsPurelyInseparable F E] [IsSeparable F E] : Function.Surjective (algebraMap F E) :=
+  fun x ↦ IsPurelyInseparable.inseparable F x (IsSeparable.separable F x)
+
+/-- If `E / F` is both purely inseparable and separable, then `algebraMap F E` is bijective. -/
+theorem IsPurelyInseparable.bijective_algebraMap_of_isSeparable
+    [IsPurelyInseparable F E] [IsSeparable F E] : Function.Bijective (algebraMap F E) :=
+  ⟨(algebraMap F E).injective, surjective_algebraMap_of_isSeparable F E⟩
+
+/-- If an intermediate field of `E / F` is both purely inseparable and separable, then it is equal
+to `F`. -/
+theorem IntermediateField.eq_bot_of_isPurelyInseparable_of_isSeparable (L : IntermediateField F E)
+    [IsPurelyInseparable F L] [IsSeparable F L] : L = ⊥ := bot_unique fun x hx ↦ by
+  obtain ⟨y, hy⟩ := IsPurelyInseparable.surjective_algebraMap_of_isSeparable F L ⟨x, hx⟩
+  exact ⟨y, congr_arg (algebraMap L E) hy⟩
+
+/-- If `E / F` is purely inseparable, then the (relative) separable closure of `E / F` is
+equal to `F`. -/
+theorem separableClosure.eq_bot_of_isPurelyInseparable [IsPurelyInseparable F E] :
+    separableClosure F E = ⊥ :=
+  bot_unique fun x h ↦ IsPurelyInseparable.inseparable F x (mem_separableClosure_iff.1 h)
+
+/-- If `E / F` is an algebraic extension, then the (relative) separable closure of `E / F` is
+equal to `F` if and only if `E / F` is purely inseparable. -/
+theorem separableClosure.eq_bot_iff (halg : Algebra.IsAlgebraic F E) :
+    separableClosure F E = ⊥ ↔ IsPurelyInseparable F E :=
+  ⟨fun h ↦ isPurelyInseparable_iff.2 fun x ↦ have hx := (halg x).isIntegral; ⟨hx, fun hs ↦ by
+    simpa only [h] using mem_separableClosure_iff.2 hs⟩,
+      fun _ ↦ eq_bot_of_isPurelyInseparable F E⟩
+
+instance isPurelyInseparable_self : IsPurelyInseparable F F :=
+  ⟨fun _ ↦ isIntegral_algebraMap, fun x _ ↦ ⟨x, rfl⟩⟩
+
+/-- A field extension `E / F` of exponential characteristic `q` is purely inseparable
+if and only if for every element `x` of `E`, there exists a natural number `n` such that
+`x ^ (q ^ n)` is contained in `F`. -/
+theorem isPurelyInseparable_iff_mem_pow (q : ℕ) [hF : ExpChar F q] :
```

Comment:
I think we should have the forward direction as its own lemma with `IsPurelyInseparable` as a typeclass variable

Reply:
Add as `IsPurelyInseparable.mem_pow`.

Reply:
> Why is the name not `pow_mem` but `mem_pow`?

Sorry, will rename it soon.
 
> Also, I feel that it would make things easier if you first define the relative perfect closure in the following way:
> 
> ```lean
> def perfectClosure : IntermediateField F E :=
>   let ⟨q, _inst⟩ := Classical.indefiniteDescription _ (ExpChar.exists F)
>   { carrier := {x : E | ∃ n : ℕ, x ^ q ^ n ∈ (algebraMap F E).range}
>     zero_mem' := ⟨0, by rw [pow_zero, pow_one]; exact zero_mem _⟩
>     one_mem' := ⟨0, by rw [pow_zero, pow_one]; exact one_mem _⟩
>     add_mem' := by
>       rintro x y ⟨n, hx⟩ ⟨m, hy⟩
>       use n + m
>       have := expChar_of_injective_algebraMap (algebraMap F E).injective q
>       rw [add_pow_expChar_pow, pow_add, pow_mul, mul_comm (q ^ n), pow_mul]
>       exact add_mem (pow_mem hx _) (pow_mem hy _)
>     mul_mem' := by
>       rintro x y ⟨n, hx⟩ ⟨m, hy⟩
>       use n + m
>       rw [mul_pow, pow_add, pow_mul, mul_comm (q ^ n), pow_mul]
>       exact mul_mem (pow_mem hx _) (pow_mem hy _)
>     inv_mem' := by
>       rintro x ⟨n, hx⟩
>       use n; rw [inv_pow]
>       apply inv_mem (id hx : _ ∈ (⊥ : IntermediateField F E))
>     algebraMap_mem' := fun x ↦ ⟨0, by rw [pow_zero, pow_one]; exact ⟨x, rfl⟩⟩ }
> ```
> 
> and rephrase this lemma as `IsPurelyInseparable F E ↔ perfectClosure F E = ⊤`. For an intermediate field `L`, you can deduce that `IsPurelyInseparable F L ↔ L ≤ perfectClosure F E`. This should make many theorems trivial (starting from `isPurelyInseparable_adjoin_finset_of_mem_pow`), and you could follow similar arguments for the "dual" (separable) case once you know `separableClosure` is a field.

Currently the `perfectClosure` is defined in a downsteram file. Let me study later to see that if it can be moved into this file.

Reply:
Your definition of perfectClosure (as the iSup of purely inseparable subextensions) requires IsPurelyInseparable, while mine doesn't. I originally thought yours is necessary because mine depends on the characteristic, until I realized yesterday that we can just choose an expChar, so we don't need to carry (q) [expChar F q] around ... we need to prove expChar is unique though.

Reply:
> Your definition of perfectClosure (as the iSup of purely inseparable subextensions) requires IsPurelyInseparable, while mine doesn't.

Yes. These two are equivalent definitions I think.

> I originally thought yours is necessary because mine depends on the characteristic, until I realized yesterday that we can just choose an expChar, so we don't need to carry (q) [expChar F q] around ...

Another idea is to use `ringExpChar R := if ringChar R = 0 then 1 else ringChar R` instead of the `q` in the definition of relative perfect closure, and in the proof of `add_mem'`, etc., we can show that if `ExpChar F q` then `q` is equal to `ringExpChar R`.

> we need to prove expChar is unique though.

For `CharP` there is [CharP.exists_unique](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/CharP/Basic.html#CharP.exists_unique). I didn't realize the importance of it, so I haven't added the corresponding version for `ExpChar`.

Reply:
or just use `max (ringChar R) 1` ...

Reply:
`ringChar.charP` is an instance which states `CharP R (ringChar R)` so I assume that `(a+b)^ringChar = a^ringChar + b^ringChar` can be rewritten automatically as usual.

Reply:
That's good to know! I guess we could add ringExpChar as an abbrev and make ExpChar R (ringExpChar R) an instance as well.